### DISCUSSION TITLE
How to use multi-threading in python with `.sql()`

### DISCUSSION THREAD

**Discussion by erezinman** [duckdb/duckdb#14369] [Q&A]
Based on [this article](https://duckdb.org/docs/guides/python/multiple_threads), I tried to use the following syntax for multi-threading:

```
base_connection = ...
connection = base_connection.cursor()
connection.sql("""SELECT 1""")
```
> duckdb.duckdb.ConnectionException: Connection Error: Connection has already been closed

However, 

```
base_connection = ...
connection = base_connection.cursor()
connection.execute("""SELECT 1""").fetchall()
```
> [(1,)]

I wish to use the `sql()` syntax or similar, because I wish to have a polars data-frame, and doing it via python should result in a substantial performance issue. 

Any ideas?

**Comment by Hunknown** [comment]
I had this error but it was because I was wrapping the call in a function, and it seems the cursor connection can only exist within the local scope, it gets closed once you get out of the function.

My function was something like:

```python
import duckdb

con = duckdb.connect()

def concurrent(query):
     with con.cursor() as cursor:
         return cursor.sql(query)

def df(query):
  return concurrent(query).df()

df("select 1")
``` 
> duckdb.duckdb.ConnectionException: Connection Error: Connection has already been closed

I fixed it by passing a cursor to the `concurrent` function:
```python
import duckdb

con = duckdb.connect()

def concurrent(query, cursor):
     return cursor.sql(query)

def df(query):
   with con.cursor() as cursor:
      return concurrent(query, cursor).df()

df("select 1")
``` 
```
	1
0	1
```

Not sure how exactly you called the cursor but hopefully this can help.
